File: boardgametracker.client/src/models/GameNight/GameNight.ts
import type { Game } from "../Games/Game";
import type { Location } from "../Location/Location";
import type { Player } from "../Player/Player";

export enum GameNightRsvpState {
  Accepted = "accepted",
  Declined = "declined",
  Pending = "pending",
}

export interface GameNightRsvps {
  id: number;
  playerId: number;
  player: Player;
  gameNightId: string;
  state: GameNightRsvpState;
}

export interface GameNight {
  id: number;
  title: string;
  startDate: Date;
  location: Location;
  locationId: number;
  host: Player;
  hostId: number;
  linkId: string;
  invitedPlayers: GameNightRsvps[];
  suggestedGames: Game[];
  notes: string;
}

export interface CreateGameNight {
  title: string;
  notes: string;
  startDate: Date;
  hostId: number;
  locationId: number;
  suggestedGameIds: number[];
  invitedPlayerIds: number[];
}

export interface UpdateGameNightRsvp {
  gameNightId: number;
  playerId: number;
  state: GameNightRsvpState;
}

export interface GameNightStatistics {
  upcomingCount: number;
  pendingResponses: number;
  gamesPlanned: number;
}

File: BoardGameTracker.Common/Entities/GameNight.cs
using BoardGameTracker.Common.Entities.Helpers;
using BoardGameTracker.Common.Enums;

namespace BoardGameTracker.Common.Entities;

public class GameNight: HasId
{
    public string Title { get; private set; } = string.Empty;
    public string Notes { get; private set; } = string.Empty;
    public Guid LinkId { get; private set; }
    public DateTime StartDate { get; private set; }

    public int HostId { get; private set; }
    public Player Host { get; private set; } = null!;
    
    public int LocationId { get; private set; }
    public Location Location { get; private set; } = null!;

    public ICollection<Game> SuggestedGames { get; private set; } = [];
    public ICollection<GameNightRsvp> InvitedPlayers { get; private set; } = [];

    public static GameNight Create(string title, string notes, DateTime startDate, int hostId, int locationId)
    {
        return new GameNight()
        {
            Title = title,
            Notes =  notes,
            StartDate = startDate,
            HostId = hostId,
            LocationId = locationId,
            LinkId = Guid.NewGuid(),
        };
    }

    public void Update(string title, string notes, DateTime startDate, int hostId, int locationId)
    {
        Title = title;
        Notes = notes;
        StartDate = startDate;
        HostId = hostId;
        LocationId = locationId;
    }

    public void SetInvitedPlayers(IEnumerable<GameNightRsvp> gameNightRsvps)
    {
        InvitedPlayers = new List<GameNightRsvp>();
        foreach (var gameNightRsvp in gameNightRsvps)
        {
            if (InvitedPlayers.Any(x => x.PlayerId == gameNightRsvp.PlayerId))
            {
                continue;
            }
        
            InvitedPlayers.Add(gameNightRsvp);
        }
    }

    public void RemoveInvitedPlayers(IEnumerable<GameNightRsvp> gameNightRsvps)
    {
        foreach (var rsvp in gameNightRsvps)
        {
            InvitedPlayers.Remove(rsvp);
        }
    }

    public void AddInvitedPlayers(IEnumerable<int> playerIds)
    {
        foreach (var playerId in playerIds)
        {
            if (InvitedPlayers.Any(x => x.PlayerId == playerId))
            {
                continue;
            }
            
            InvitedPlayers.Add(GameNightRsvp.Create(playerId, GameNightRsvpState.Pending));
        }
    }

    public void SetSuggestedGames(IEnumerable<Game> suggestedGames)
    {
        SuggestedGames = new List<Game>();
        foreach (var suggestedGame in suggestedGames)
        {
            if (SuggestedGames.Any(x => x.Id == suggestedGame.Id))
            {
                continue;
            }
        
            SuggestedGames.Add(suggestedGame);
        }
    }
}